projects/packages/image-cdn/src/class-image-cdn.php

```diff
@@ -0,0 +1,1386 @@
+	 *
+	 * @return array Empty array.
+	 */
+	public static function filter_photon_noresize_intermediate_sizes() {
+		return array();
+	}
+
+	/**
+	 * Filter thumbnail URLS to not generate.
+	 *
+	 * @param array $sizes Image sizes.
+	 *
+	 * @return mixed
+	 */
+	public static function filter_photon_noresize_thumbnail_urls( $sizes ) {
+		foreach ( $sizes as $size => $url ) {
+			$parts     = explode( '?', $url );
+			$arguments = isset( $parts[1] ) ? $parts[1] : array();
+
+			$sizes[ $size ] = Image_CDN_Core::cdn_url( $url, wp_parse_args( $arguments ) );
+		}
+
+		return $sizes;
+	}
+
+	/**
+	 * Inject image sizes to attachment metadata.
+	 *
+	 * @param array $data          Attachment metadata.
+	 * @param int   $attachment_id Attachment's post ID.
+	 *
+	 * @return array Attachment metadata.
+	 */
+	public static function filter_photon_norezise_maybe_inject_sizes( $data, $attachment_id ) {
+		// Can't do much if data is empty.
+		if ( empty( $data ) ) {
+			return $data;
+		}
+		$sizes_already_exist = (
+			true === is_array( $data )
+			&& true === array_key_exists( 'sizes', $data )
+			&& true === is_array( $data['sizes'] )
+			&& false === empty( $data['sizes'] )
+		);
+		if ( $sizes_already_exist ) {
+			return $data;
+		}
+		// Missing some critical data we need to determine sizes, not processing.
+		if ( ! isset( $data['file'] )
+			|| ! isset( $data['width'] )
+			|| ! isset( $data['height'] )
+		) {
+			return $data;
+		}
+
+		$mime_type           = get_post_mime_type( $attachment_id );
+		$attachment_is_image = preg_match( '!^image/!', $mime_type );
+
+		if ( 1 === $attachment_is_image ) {
+			$image_sizes   = new Image_CDN_Image_Sizes( $attachment_id, $data );
+			$data['sizes'] = $image_sizes->generate_sizes_meta();
+		}
+		return $data;
+	}
+
+	/**
+	 * Inject image sizes to Jetpack REST API responses. This wraps the filter_photon_norezise_maybe_inject_sizes function.
+	 *
+	 * @param array $sizes Attachment sizes data.
+	 * @param int   $attachment_id Attachment's post ID.
+	 *
+	 * @return array Attachment sizes array.
+	 */
+	public static function filter_photon_norezise_maybe_inject_sizes_api( $sizes, $attachment_id ) {
+		return self::filter_photon_norezise_maybe_inject_sizes( wp_get_attachment_metadata( $attachment_id ), $attachment_id );
+	}
+
+	/**
+	 * * IN-CONTENT IMAGE MANIPULATION FUNCTIONS
+	 **/
+
+	/**
+	 * Match all images and any relevant <a> tags in a block of HTML.
+	 *
+	 * @param string $content Some HTML.
+	 * @return array An array of $images matches, where $images[0] is
+	 *         an array of full matches, and the link_url, img_tag,
+	 *         and img_url keys are arrays of those matches.
+	 */
+	public static function parse_images_from_html( $content ) {
+		$images = array();
+
+		if ( preg_match_all( '#(?:<a[^>]+?href=["|\'](?P<link_url>[^\s]+?)["|\'][^>]*?>\s*)?(?P<img_tag><(?:img|amp-img|amp-anim)[^>]*?\s+?src=["|\'](?P<img_url>[^\s]+?)["|\'].*?>){1}(?:\s*</a>)?#is', $content, $images ) ) {
```

Comment:
Pretty sure there's some typos in this regex
```suggestion
		if ( preg_match_all( '#(?:<a[^>]+?\s+?href=["\'](?P<link_url>[^\s]+?)["\'][^>]*?>\s*)?(?P<img_tag><(?:img|amp-img|amp-anim)[^>]*?\s+?src=["\'](?P<img_url>[^\s]+?)["\'].*?>){1}(?:\s*</a>)?#is', $content, $images ) ) {
```
Or do we really want to be looking for `<a href=|url|>`?

Reply:
Good eye. Your suggested update requires two characters between `<a` and `href` though, so `<a href='cheese"><img src='x'></a>` doesn't match.

I've gone ahead with `[^>]*?\s+?` in be6dfd69f6 to allow 0 characters before the space before href.

Reply:
Oops, good catch!